man/man8/dslower.8
.TH dslower 8  "2022-12-11" "USER COMMANDS"
.SH NAME
dslower \- Trace long process block time(in D state). Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B dslower [\-h] [\-p PID | \-t TID | \-c PPID] [\-s] [min_us]
.SH DESCRIPTION
This script traces long process block time(in D state).
This measures the time between a task was switched off cpu because of
uninterruptible sleep and been woken up. ie: sched_switch -> ttwu_do_wakeup

Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-h
Print usage message.
.TP
\-p PID
Trace this process ID only (filtered in-kernel).
.TP
\-t TID
Trace this thread ID only (filtered in-kernel).
.TP
\-c PPID
Trace only child of this pid.
.TP
\-s
also show stack trace
.SH EXAMPLES
.TP
Trace block time longer than 10ms for all process
#
.B dslower
.TP
Trace only  PID 181:
#
.B dslower \-p 181
.TP
show stack
#
.B dslower \-s
.SH FIELDS
.TP
TIME
Time of the event.
.TP
CPU
cpu id the event occur
.TP
TID
Thread ID
.TP
COMM
Process name for the PID.
.TP
LAT
block time in us
.SH OVERHEAD
This traces scheduler functions, which can become very frequent. While eBPF
has very low overhead, and this tool uses in-kernel maps for efficiency, the
frequency of scheduler events for some workloads may be high enough that the
overhead of this tool becomes significant. Measure in a lab environment
to quantify the overhead before use.
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Curu Wong
.SH SEE ALSO
runqslower(8)


debian/changelog
bcc (0.29.0-1) unstable; urgency=low

  * Support for kernel up to 6.6.
  * new bcc tools: rdmaucma
  * new libbpf tools: f2fs, futexctn
  * bcc tool update: tcpstates, statsnoop, runqlat, bio tools, tcptop, slabratetop, tcprtt, etc.
  * libbpf tool update: tcprtt, tcppktlat, bio tools, execsnoop, bindsnoop, exitsnoop, etc.
  * s390x support for libbpf-tools
  * examples for perf/ipc
  * expose pid parameter in bpf_open_perf_event
  * allow for installing python as a non-system package
  * ci improvement: deprecate ubuntu 18.04 and allow multiple llvm versions
  * ci improvement: reitre fedora 34/26 and use fedora 38.
  * fix misaligned pointer accesses in some ringbuf using libbpf-tools
  * some new  enhancement for powerpc and riscv.
  * consolidate tools into different categories: filesystem/storage, networking, cpu and scheduling, etc.
  * doc update, other bug fixes and tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Thu, 6 Dec 2023 17:00:00 +0000

bcc (0.28.0-1) unstable; urgency=low

  * Support for kernel up to 6.3.
  * new libbpf tool: tcppktlat.
  * bcc tool updates: funcslower, wakeuptime, profile, offcputime, deadlock, funccount, argdist, kvmexit, runqlen and cpuunclaimed.
  * libbpf tool update: memleak, tcprtt, tcpconnlat, funclatency, syscount, cpufreq, biosnoop.
  * support ringbuf_query for bcc tools.
  * handle '[uprobes]' memory mapped file properly during stack tracing.
  * Fix maximum allowed index for print_linear_hist for bcc tools.
  * add module kfunc/kretfunc support.
  * clang rewriter: initialize only the requested parameters
  * filter with available_filter_functions to make multi-functions kprobes more robust for both bcc and libbpf tools.
  * doc update, other bug fixes and tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 28 Jun 2023 17:00:00 +0000

bcc (0.27.0-1) unstable; urgency=low

  * Support for kernel up to 6.2
  * bcc tool updates for ttysnoop, slabratetop, readahead, nfsslower, cpudist, cachetop, cachestat, etc.
  * libbpf-tools updates for mdflush, drsnoop, statsnoop, ttysnoop, softirqs, wakeuptime, cachestat, numamove, etc.
  * fix for incomplete static libraries
  * implement zip archive support
  * upgrade to use c++14 standard
  * new libbpf-tools: memleak
  * add loongarch support in libbpft-tools
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 02 Apr 2023 17:00:00 +0000

bcc (0.26.0-1) unstable; urgency=low

  * Support for kernel up to 6.1
  * bcc tool updates for biosnoop, opensnoop, biopattern, killsnoop, runqslower, offcputime, wakeuptime, etc.
  * libbpf-tools updates for klockstat, sigsnoop, hardirqs, softirqs, opensnoop, statsnoop, offcputime, tcplife, cpufreq, cpudist, etc.
  * new libbpf-tools: tcptop, tcpstates, biotop, capable
  * ci: add support for fedora 36 container and new workflow for containers
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 10 Aug 2022 17:00:00 +0000

bcc (0.25.0-1) unstable; urgency=low

  * Support for kernel up to 5.19
  * bcc tool updates for oomkill.py, biolatpcts.py, sslsniff.py, tcpaccept.py, etc.
  * libbpf tool updates for klockstat, opensnoop, tcpconnect, etc.
  * new bcc tools: tcpcong
  * new libbpf tools: tcpsynbl, mdflush, oomkill, sigsnoop
  * usdt: support xmm registers as args for x64
  * bpftool as a submodule now
  * remove uses of libbpf deprecated APIs
  * use new llvm pass manager
  * support cgroup filtering libbpf tools
  * fix shared lib module offset <-> global addr conversion
  * riscv support
  * LoongArch support
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 10 Aug 2022 17:00:00 +0000

bcc (0.24.0-1) unstable; urgency=low

  * Support for kernel up to 5.16
  * bcc tools: update for trace.py, sslsniff.py, tcptop.py, hardirqs.py, etc.
  * new libbpf tools: bashreadline
  * allow specify wakeup_events for perf buffer
  * support BPF_MAP_TYPE_{INODE, TASK}_STORAGE maps
  * remove all deprecated libbpf function usage
  * remove P4/B language support
  * major test infra change, using github actions now
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 14 Jan 2022 17:00:00 +0000

bcc (0.23.0-1) unstable; urgency=low

  * Support for kernel up to 5.15
  * bcc tools: update for kvmexit.py, tcpv4connect.py, cachetop.py, cachestat.py, etc.
  * libbpf tools: update for update for mountsnoop, ksnoop, gethostlatency, etc.
  * fix renaming of task_struct->state
  * get pid namespace properly for a number of tools
  * initial work for more libbpf utilization (less section names)
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Nov 2021 17:00:00 +0000

bcc (0.22.0-1) unstable; urgency=low

  * Support for kernel up to 5.14
  * add ipv4/ipv6 filter support for tcp trace tools
  * add python interface to attach raw perf events
  * fix tcpstates for incorrect display of dport
  * new options for bcc tools runqslower, argdist
  * new libbpf-tools: filetop, exitsnoop, tcprtt
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Sep 2021 17:00:00 +0000

bcc (0.21.0-1) unstable; urgency=low

  * Support for kernel up to 5.13
  * support for debug information from libdebuginfod
  * finished support for map elements items_*_batch() APIs
  * add atomic_increment() API
  * support attach_func() and detach_func() in python
  * fix displaying PID instead of TID for many tools
  * new tools: kvmexit.py
  * new libbpf-tools: gethostlatency, statsnoop, fsdist and solisten
  * fix tools ttysnoop/readahead for newer kernels
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 16 Jul 2021 17:00:00 +0000

bcc (0.20.0-1) unstable; urgency=low

  * Support for kernel up to 5.12
  * Some basic support for MIPS

tools/dslower.py
#!/usr/bin/env python
# @lint-avoid-python-3-compatibility-imports
#
# dslower    Trace process block on D state time. (uninterruptible sleep)
#               For Linux, uses BCC, eBPF.
#
# This script traces long process block time(in D state)
#
# USAGE: dslower [-p PID] [-t TID] [-P] [min_us]
#
# REQUIRES: Linux 4.9+ (BPF_PROG_TYPE_PERF_EVENT support).
#
# This measures the time between a task was switched off cpu because of uninterruptible sleep and 
# been woken up.
# ie. sched_switch -> ttwu_do_wakeup
# Copyright 2022 Tencent
# Licensed under the Apache License, Version 2.0 (the "License")
#
# 28-Nov-2022   Curu Wong

from __future__ import print_function
from bcc import BPF
import argparse
from time import strftime

# arguments
examples = """examples:
    ./dslower         # trace process block longer than 10000 us (default)
    ./dslower 1000    # trace process block longer than 1000 us
    ./dslower -p 123  # trace pid 123
    ./dslower -t 123  # trace tid 123 (use for threads only)
    ./dslower -c 123  # trace only child of pid 123
    ./dslower -s  # also show stack trace

"""
parser = argparse.ArgumentParser(
    description="Trace process block on D state time. (uninterruptible sleep)",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("min_us", nargs="?", default='10000',
    help="minimum block time to trace, in us (default 10000)")
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)

thread_group = parser.add_mutually_exclusive_group()
thread_group.add_argument("-p", "--pid", metavar="PID", dest="pid",
    help="trace this PID only", type=int)
thread_group.add_argument("-t", "--tid", metavar="TID", dest="tid",
    help="trace this TID only", type=int)
thread_group.add_argument("-c", "--ppid", type=int,
    help="trace only process whose ppid is PPID")
parser.add_argument("--state", choices=["S", "D", ""], default="D",
    help="filter task state(default D)")
parser.add_argument("-s", "--stack", action="store_true",
    help="also show block stack trace")

args = parser.parse_args()

min_us = int(args.min_us)
debug = 0

# define BPF program
bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
#include <linux/nsproxy.h>
#include <linux/pid_namespace.h>

#ifndef TASK_NOLOAD
#define TASK_NOLOAD         0x0400
#endif

struct ts_stack_t {
    u32 stack_id;
    u64 ts;
};
BPF_ARRAY(start, struct ts_stack_t, MAX_PID);

struct rq;

BPF_STACK_TRACE(stack_traces, 4096);

struct data_t {
    u32 pid;
    u32 stack_id;
    char task[TASK_COMM_LEN];
    u64 delta_us;
};

BPF_PERF_OUTPUT(events);
"""

bpf_text_kprobe = """
TRACE_FINISH_TASK_SWITCH_FUNC
{
    u32 pid, tgid, ppid;
    u64 ts = bpf_ktime_get_ns();

    // prev task go to sleep
    long state = prev->STATE_FIELD;
    if (state != 0) {
        if(STATE_FILTER) {
            pid = prev->pid;
            tgid = prev->tgid;
            ppid = prev->real_parent->tgid;
            if (pid != 0) {
                if (!(FILTER_PID) && !(FILTER_TGID) && !(FILTER_PPID)) {
                    struct ts_stack_t ts_stack = { .ts = ts };
                    start.update(&pid, &ts_stack);
                }
            }
        }
    }

    /* finish_task_switch already switched stack, so we can't get stack trace of prev task
     *  we need to get the stack after the same task sched in again.
     *  the accurate delta should between sched_swith -> sched_wakeup
     *  the delta caculated here is longer than real block time, this is indeed  block_time + run_delay
    */
    pid = bpf_get_current_pid_tgid();
    struct ts_stack_t *ts_stack_p = start.lookup(&pid);

    u64 delta_us;
    if ((ts_stack_p == 0) || (ts_stack_p->ts == 0)) {
        return 0;   // missed enqueue
    }

    if(ts < ts_stack_p->ts){
        //maybe time wrap
        ts_stack_p->ts = 0;
        return 0;
    }

    delta_us = (ts - ts_stack_p->ts) / 1000;
    ts_stack_p->ts = 0;

    if (FILTER_US){
        return 0;
    }

    struct data_t data = {};
    data.pid = pid;
    data.delta_us = delta_us;
    bpf_get_current_comm(&data.task, sizeof(data.task));
#ifdef SHOW_STACK
    data.stack_id = stack_traces.get_stackid(ctx, 0);
#endif

    // output
    events.perf_submit(ctx, &data, sizeof(data));

    return 0;

}
"""

bpf_text_raw_tp = """
RAW_TRACEPOINT_PROBE(sched_wakeup)
{
    // TP_PROTO(struct task_struct *p)
    struct task_struct *p = (struct task_struct *)ctx->args[0];

    u64 *tsp, delta_us;
    u32 pid = p->pid;
    u32 tgid = p->tgid;
    u32 ppid = p->real_parent->tgid;

    if (FILTER_PID || FILTER_TGID || FILTER_PPID || pid == 0)
        return 0;

    u64 ts = bpf_ktime_get_ns();

    // fetch timestamp and calculate delta
    struct ts_stack_t *ts_stack_p = start.lookup(&pid);
    if ((ts_stack_p == 0) || (ts_stack_p->ts == 0)) {
        return 0;   // missed enqueue
    }

    if(ts < ts_stack_p->ts){
        //maybe time wrap
        ts_stack_p->ts = 0;
        return 0;
    }

    delta_us = (ts - ts_stack_p->ts) / 1000;
    ts_stack_p->ts = 0;

    if (FILTER_US){
        return 0;
    }

    struct data_t data = {};
    data.pid = pid;
    data.delta_us = delta_us;
    bpf_probe_read(&data.task, sizeof(data.task), p->comm);
    data.stack_id = ts_stack_p->stack_id;

    // output
    events.perf_submit(ctx, &data, sizeof(data));

    return 0;
}

RAW_TRACEPOINT_PROBE(sched_switch)
{
    // TP_PROTO(bool preempt, struct task_struct *prev, struct task_struct *next)
    struct task_struct *prev = (struct task_struct *)ctx->args[1];

    u32 pid, tgid, ppid;

    // prev task go to sleep
    long state = prev->STATE_FIELD;
    if ( state != 0 ) {
        if(STATE_FILTER){
            pid = prev->pid;
            tgid = prev->tgid;
            ppid = prev->real_parent->tgid;
            u64 ts = bpf_ktime_get_ns();
            if (pid != 0) {
                if (!(FILTER_PID) && !(FILTER_TGID) && !(FILTER_PPID)) {
                    struct ts_stack_t ts_stack = { .ts = ts };
#ifdef SHOW_STACK
                    ts_stack.stack_id = stack_traces.get_stackid(ctx, 0);
#endif
                    start.update(&pid, &ts_stack);
                }
            }
        }
    }
    return 0;
}
"""

is_support_raw_tp = BPF.support_raw_tracepoint()
if is_support_raw_tp:
    bpf_text += bpf_text_raw_tp
else:
    bpf_text += bpf_text_kprobe

# code substitutions
if BPF.kernel_struct_has_field(b'task_struct', b'__state') == 1:
    bpf_text = bpf_text.replace('STATE_FIELD', '__state')
else:
    bpf_text = bpf_text.replace('STATE_FIELD', 'state')

bpf_text_finish_task_switch_old = """
int trace_finish_task_switch(struct pt_regs *ctx, struct rq *rq, struct task_struct *prev)
"""
bpf_text_finish_task_switch_new = """
int trace_finish_task_switch(struct pt_regs *ctx, struct task_struct *prev)
"""
# kernel v3.19 task_struct added numa_faults field and changed proto of finish_task_switch
if BPF.kernel_struct_has_field(b'task_struct', b'numa_faults') == 1:
    bpf_text = bpf_text.replace('TRACE_FINISH_TASK_SWITCH_FUNC', bpf_text_finish_task_switch_new)
else:
    bpf_text = bpf_text.replace('TRACE_FINISH_TASK_SWITCH_FUNC', bpf_text_finish_task_switch_old)

if min_us == 0:
    bpf_text = bpf_text.replace('FILTER_US', '0')
else:
    bpf_text = bpf_text.replace('FILTER_US', 'delta_us <= %s' % str(min_us))

if args.tid:
    bpf_text = bpf_text.replace('FILTER_PID', 'pid != %s' % args.tid)
else:
    bpf_text = bpf_text.replace('FILTER_PID', '0')

if args.pid:
    bpf_text = bpf_text.replace('FILTER_TGID', 'tgid != %s' % args.pid)
else:
    bpf_text = bpf_text.replace('FILTER_TGID', '0')
if args.ppid:
    bpf_text = bpf_text.replace('FILTER_PPID', 'ppid != %s' % args.ppid)
else:
    bpf_text = bpf_text.replace('FILTER_PPID', '0')

if args.state == 'S':
    bpf_text = bpf_text.replace('STATE_FILTER', 'state & TASK_INTERRUPTIBLE')
elif args.state == 'D':
    bpf_text = bpf_text.replace('STATE_FILTER', 
        '(state & TASK_UNINTERRUPTIBLE) && !(state & TASK_NOLOAD) ')
else:
    bpf_text = bpf_text.replace('STATE_FILTER', '0');


if debug or args.ebpf:
    print(bpf_text)
    if args.ebpf:
        exit()

# process event
def print_event(cpu, data, size):
    event = b["events"].event(data)
    print("%-8s [%03d] %-16s %-7s %d" % (strftime("%H:%M:%S"), cpu, event.task.decode('utf-8', 'replace'), event.pid, event.delta_us))
    if args.stack:
        for addr in stack_traces.walk(event.stack_id):
            sym = b.ksym(addr, show_offset=True).decode('utf-8', 'replace')
            print("\t%s" % sym)

max_pid = int(open("/proc/sys/kernel/pid_max").read())

# load BPF program
cflags = ["-DMAX_PID=%d" % max_pid]
if args.stack:
    cflags.append("-DSHOW_STACK")
b = BPF(text=bpf_text, cflags=cflags)
if not is_support_raw_tp:
    b.attach_kprobe(event_re="^finish_task_switch$|^finish_task_switch\.isra\.\d$",
                    fn_name="trace_finish_task_switch")
print("Tracing block time longer higher than %d us" % min_us)
print("%-8s %-5s %-16s %-7s %s" % ("TIME","CPU", "COMM", "TID", "LAT(us)"))

stack_traces = b.get_table("stack_traces")

# read events
b["events"].open_perf_buffer(print_event, page_cnt=64)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
